src/tpm2/crypto/openssl/CryptSym.c

```diff
@@ -458,6 +461,262 @@ CryptSymmetricDecrypt(
+		      INT32                dSize,         // IN: data size (may need to be a
+		      //     multiple of the blockSize)
+		      const BYTE          *dIn            // IN: data buffer
+		      )
+{
+    INT16                blockSize;
+    BYTE                *iv;
+    BYTE                 defaultIv[MAX_SYM_BLOCK_SIZE] = {0};
+    evpfunc              evpfn;
+    EVP_CIPHER_CTX      *ctx = NULL;
+    int                  outlen1, outlen2;
+    BYTE                *pOut = dOut;
+    BYTE                *buffer = NULL; // for in-place encryption
+    UINT32               buffersize = 0;
+    BYTE                 keyToUse[MAX_SYM_KEY_BYTES];
+    UINT16               keyToUseLen = (UINT16)sizeof(keyToUse);
+    TPM_RC               retVal = TPM_RC_SUCCESS;
+
+    pAssert(dOut != NULL && key != NULL && dIn != NULL);
+    if(dSize == 0)
+	return TPM_RC_SUCCESS;
+    TEST(algorithm);
+    blockSize = CryptGetSymmetricBlockSize(algorithm, keySizeInBits);
+    if(blockSize == 0)
+	return TPM_RC_FAILURE;
+    // If the iv is provided, then it is expected to be block sized. In some cases,
+    // the caller is providing an array of 0's that is equal to [MAX_SYM_BLOCK_SIZE]
+    // with no knowledge of the actual block size. This function will set it.
+    if((ivInOut != NULL) && (mode != ALG_ECB_VALUE))
+	{
+	    ivInOut->t.size = blockSize;
+	    iv = ivInOut->t.buffer;
+	}
+    else
+	iv = defaultIv;
+
+    switch (mode)
+        {
+          case ALG_ECB_VALUE:
+          case ALG_CBC_VALUE:
+	    // For ECB & CBC the data size must be an even multiple of the
+	    // cipher block size
+	    if((dSize % blockSize) != 0)
+		return TPM_RC_SIZE;
+        }
+
+    evpfn = GetEVPCipher(algorithm, keySizeInBits, mode, key,
+                         keyToUse, &keyToUseLen);
+    if (evpfn == NULL)
+        return TPM_RC_FAILURE;
+
+    if (dIn == dOut) {
+        // in-place encryption; we use a temp buffer
+        buffersize = TPM2_ROUNDUP(dSize, blockSize);
+        buffer = malloc(buffersize);
+        if (buffer == NULL)
+            ERROR_RETURN(TPM_RC_FAILURE);
+
+        pOut = buffer;
+    }
+
+#if ALG_TDES && ALG_CTR
+    if (algorithm == TPM_ALG_TDES && mode == ALG_CTR_VALUE) {
+        TDES_CTR(keyToUse, keyToUseLen * 8, dSize, dIn, iv, pOut, blockSize);
+        if (pOut != dOut)
+            memcpy(dOut, pOut, dSize);
+        ERROR_RETURN(TPM_RC_SUCCESS);
+    }
+#endif
+
+    ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    if (EVP_EncryptInit_ex(ctx, evpfn(), NULL, keyToUse, iv) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
+    if (EVP_EncryptUpdate(ctx, pOut, &outlen1, dIn, dSize) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
+    if (EVP_EncryptFinal_ex(ctx, pOut + outlen1, &outlen2) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    pAssert(dSize >= outlen1 + outlen2);
```

Comment:
It is too late to asser there, you may have already overwritten memory, the check, if you think is needed, should be done beofre calling EVP_EncryptFinal_ex().
Something like this I think:
pAssert(oulen1 == dSize || dSize >= outlen1 + blockSize);

Reply:
Right. Thank you.